### PR TITLE
Major Main Training Loop Rewrite

### PR DESCRIPTION
**Pull request by raunakdoesdev**
- Switched to a moment based training loop
- Added full epoch validation after each epoch of training
- Added extremely detailed logging for quick debugging
- Added record of training losses in csv file for easy data analysis

I've tested this pretty thoroughly and everything seems to be working, but I *HIGHLY* recommend this is reviewed thoroughly before merging, since it is such a huge change. In fact it would be good if multiple people could approve and check off on these changes before the merge.

### CODE REVIEW

**Comment by raunakdoesdev** [comment]
@tpankaj The 500 was a number I used for testing, nice catch. I realized I still haven't added a lot of the constants I use to the Paramters.py file. I'll do this now.

**Comment by raunakdoesdev** [comment]
@tpankaj Done

File: Train.py [hunk 1]
```diff
@@ -24,63 +29,77 @@
...
-
-            batch.fill(data, data_index)  # Get batches ready
+def run_net(data_index):
+    batch.fill(data, data_index)  # Get batches ready
+    batch.forward(optimizer, criterion, data_moment_loss_record)
 
-            # Run net, forward pass
-            batch.forward(optimizer, criterion, data_moment_loss_record)
+try:
+    for epoch in range(1000):
```
Comment (tpankaj) [inline Train.py]: This should be a parameter or at least a constant defined at the top of the file.

File: Utils.py [hunk 2]
```diff
@@ -1,8 +1,40 @@
 from Parameters import args
 from libs.utils2 import *
 import matplotlib.pyplot as plt
+import numpy
```
Comment (tpankaj) [inline Utils.py]: The NumPy convention is to do import numpy as np.

File: Utils.py [hunk 3]
```diff
@@ -1,8 +1,40 @@
...
 import torch
 
+class Moment_Counter:
+    """Notify after N Data Moments Passed"""
+    def __init__(self, n):
+        self.start = 0
+        self.n = n
+
+    def step(self, data_index):
+        if data_index.ctr - self.start > self.n:
```
Comment (tpankaj) [inline Utils.py]: Should it be > or >= ?

File: Train.py [hunk 4]
```diff
@@ -24,63 +29,77 @@
...
+
+        data.train_index.epoch_complete = False
+        epoch_train_loss.export_csv('logs/epoch%02d_train_loss.csv' % (epoch,))
+        logging.info('Avg Train Loss = {}'.format(epoch_train_loss.average()))
+        logging.debug('Finished training epoch #{}'.format(epoch))
+        logging.debug('Starting validation epoch #{}'.format(epoch))
+        epoch_val_loss = Utils.Loss_Log()
+
+        net.eval()  # Evaluate mode
+        while not data.val_index.epoch_complete:
```
Comment (tpankaj) [inline Train.py]: Progress prints during validation?

File: Train.py [hunk 5]
```diff
@@ -11,6 +11,11 @@
 from nets.SqueezeNet import SqueezeNet
 import torch
 
+import traceback
+import logging
+logging.basicConfig(filename='training.log', level=logging.DEBUG)
+logging.debug(args)  # Log arguments
```
Comment (tpankaj) [inline Train.py]: Remove trailing whitespace

File: Train.py [hunk 6]
```diff
@@ -24,63 +29,77 @@
...
+        print_counter = Utils.Moment_Counter(args.print_moments)
 
-            if mode == 'train':  # Backpropagate
-                batch.backward(optimizer)
+        while not data.train_index.epoch_complete: # Epoch of training
+            run_net(data.train_index)  # Run network
+            batch.backward(optimizer)  # Backpropagate
 
-            loss_record[mode].add(batch.loss.data[0])
+            # Logging Loss
```
Comment (tpankaj) [inline Train.py]: Remove trailing whitespace

File: Train.py [hunk 7]
```diff
@@ -24,63 +29,77 @@
...
+        data.val_index.epoch_complete = False
+        epoch_val_loss.export_csv('logs/epoch%02d_val_loss.csv' % (epoch,))
+        logging.debug('Finished validation epoch #{}'.format(epoch))
+        logging.info('Avg Val Loss = {}'.format(epoch_val_loss.average()))
+        Utils.save_net("save/epoch%02d_save_%f" % (epoch,\
+                       epoch_val_loss.average()), net)
+except Exception:
+    logging.error(traceback.format_exc())  # Log exception
+
+    # Interrupt Saves
```
Comment (tpankaj) [inline Train.py]: Remove trailing whitespace

**Review by tpankaj — CHANGES_REQUESTED** [review]
Please address all of my comments and I will review again.

**Review by tpankaj — DISMISSED** [review]

**Review by tpankaj — APPROVED** [review]

**Pull request merged**